### PR TITLE
sql: Deduplication logic for `ALTER PRIMARY KEY`

### LINKED ISSUE #74116: Deduplicate unique indexes when altering primary key
**Is your feature request related to a problem? Please describe.**
When we alter primary key, we create a unique index to match the old primary key. However, it doest that no matter if there's already a same unique key exists or not. So we have duplicate unique indexes in this case.
```sql
--(1) create table with unique index on primary key columns
create table t (a int primary key, b int not null, unique index t_a_key (a));

--(2) alter primary key
alter table t alter primary key using columns (b);

--(3) t has duplicate unique index on column (a)
show create table t;

  table_name |              create_statement
-------------+---------------------------------------------
  t          | CREATE TABLE public.t (
             |     a INT8 NOT NULL,
             |     b INT8 NOT NULL,
             |     CONSTRAINT t_pkey PRIMARY KEY (b ASC),
             |     UNIQUE INDEX t_a_key1 (a ASC),
             |     UNIQUE INDEX t_a_key (a ASC),
             |     FAMILY "primary" (a, b)
             | )
```

**Describe the solution you'd like**
We make add some deduplication here.

Jira issue: CRDB-11924

### PR DESCRIPTION
**Pull request by Xiang-Gu**
When we use `ALTER PRIMARY KEY` to change the primary key of a table, a new
secondary index on the old primary key columns will be created. 
This PR added deduplication logic for `ALTER PRIMARY KEY` if there exists a unique,
non-partial secondary index that has the same columns as the (old) PK columns in
the same order with same direction.

Release note (sql change): `ALTER PRIMARY KEY` will not create a seconday index on 
the old PK columns if there is already a unique, non-partial secondary index that is
"identical" to the (old) PK columns. Two indexes are "identical" if they have exactly
the same columns, in the same order, with same direction.

Fixes: #74116

Jira issue: CRDB-14766

### CODE REVIEW

**Review by postamar — COMMENTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78046)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Xiang-Gu)

___
*[pkg/sql/alter_primary_key.go, line 663 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhLo524ETuuCyRhA1v:-MyhLo524ETuuCyRhA1w:bm164mg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L663)):*
> ```Go
> 		return true
> 	}
> 	alreadyHasSecondaryIndexOnPKColumns := func(desc *tabledesc.Mutable) bool {
> ```

Can you rewrite this with `desc` being of type `catalog.TableDescriptor` instead? We try to use these `catalog` interfaces wherever possible. There's no way you could have guessed this so don't feel bad.

Really, this whole `shouldCopyPrimaryKey` function should be rewritten in this way, the only reason it hasn't is because it was written before these interfaces were introduced about a year ago.
___
*[pkg/sql/alter_primary_key.go, line 666 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhMrUw4NbzW6iRvEtA:-MyhMYcj5x4zbFtfDvDN:bqctolz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L666)):*
> ```Go
> 		// Return whether any existing secondary index has identical key columns
> 		// as the primary key columns.
> 		pk := desc.GetPrimaryIndex().IndexDesc()
> ```

In light of the above comment, this should become `pk := desc.GetPrimaryIndex()` which returns a `catalog.Index`. The `.IndexDesc()` method returns the underlying descriptor protobuf, this has the same information in it but is much less safe to use. Partly because the interface is immutable by design so it's less easy to overwrite a value by mistake, but mostly because we have to keep descriptor protobufs compatible across cluster versions so sometimes some fields have unintuitive values.
___
*[pkg/sql/alter_primary_key.go, line 667 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhO-d78XLuPNgb31ir:-MyhO-d78XLuPNgb31is:b-vcsr2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L667)):*
> ```Go
> 		// as the primary key columns.
> 		pk := desc.GetPrimaryIndex().IndexDesc()
> 		for _, existingIndex := range desc.GetIndexes() {
> ```

In light of the previous, here you'll want to iterate over `desc.PublicNonPrimaryIndexes()`.
___
*[pkg/sql/alter_primary_key.go, line 670 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhOFzKAaslAs6U5riF:-MyhOFzKAaslAs6U5riG:bizlzz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L670)):*
> ```Go
> 			if len(pk.KeyColumnIDs) != len(existingIndex.KeyColumnIDs) || !existingIndex.Unique {
> 				continue
> 			}
> ```

You should also check that `existingIndex` is not a partial index. It isn't the most obvious thing in the world but we allow WHERE clauses in indexes! To reuse your example in the logic test, this could be something like `CREATE TABLE t (a INT PRIMARY KEY, b INT NOT NULL, UNIQUE INDEX t_a_key (a) WHERE (a > 0));`.

You can check for the presence of this partial predicate with the `IsPartial` method on `catalog.Index`.
___
*[pkg/sql/alter_primary_key.go, line 674 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhOTTk5rTkzIOpyq3m:-MyhOTTk5rTkzIOpyq3n:b-y7ryhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L674)):*
> ```Go
> 			for i := range pk.KeyColumnIDs {
> 				if pk.KeyColumnIDs[i] != existingIndex.KeyColumnIDs[i] ||
> 					pk.KeyColumnDirections[i] != existingIndex.KeyColumnDirections[i] {
> ```

I'm not sure if column directions matter here, or elsewhere in `shouldCopyPrimaryKey` for that matter. Let's ask @ajwerner 
___
*[pkg/sql/alter_primary_key.go, line 678 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhOnr82RMSA560QotY:-MyhOnr82RMSA560QotZ:b-qxcmcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L678)):*
> ```Go
> 					break
> 				}
> 			}
> ```

`catalog.Index` has a nice `CollectKeyColumnIDs` method which returns a `catalog.TableColSet` which should be much more convenient for what you want to do here, which is to test whether there exists a non-partial unique secondary index whose set of key columns is a superset of the PK columns.
___
*[pkg/sql/logictest/testdata/logic_test/alter_primary_key, line 1607 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhKb7m2mJQf3D_gCKi:-MyhKb7m2mJQf3D_gCKj:b5gdwvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L1607)):*
> ```
> t_pkey     a   N/A
> t_a_key    a   ASC
> t_a_key    b   ASC
> ```

This is a good test. Perhaps you can also add a case for `CREATE TABLE t (a INT PRIMARY KEY, b INT NOT NULL, UNIQUE INDEX idx (a, b));`, we should see a similar outcome: only 2 indexes (the old primary index and `idx`), the old primary index should be discarded.

**Review by Xiang-Gu — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78046)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/sql/alter_primary_key.go, line 663 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhLo524ETuuCyRhA1v:-MyhZiwi15XK6U4tD9Pk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L663)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Can you rewrite this with `desc` being of type `catalog.TableDescriptor` instead? We try to use these `catalog` interfaces wherever possible. There's no way you could have guessed this so don't feel bad.

Really, this whole `shouldCopyPrimaryKey` function should be rewritten in this way, the only reason it hasn't is because it was written before these interfaces were introduced about a year ago.
</blockquote></details>

Done.
___
*[pkg/sql/alter_primary_key.go, line 666 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhMrUw4NbzW6iRvEtA:-MyhZjWA9RR1avQ6k3OQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L666)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

In light of the above comment, this should become `pk := desc.GetPrimaryIndex()` which returns a `catalog.Index`. The `.IndexDesc()` method returns the underlying descriptor protobuf, this has the same information in it but is much less safe to use. Partly because the interface is immutable by design so it's less easy to overwrite a value by mistake, but mostly because we have to keep descriptor protobufs compatible across cluster versions so sometimes some fields have unintuitive values.
</blockquote></details>

Done.
___
*[pkg/sql/alter_primary_key.go, line 667 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhO-d78XLuPNgb31ir:-MyhZvDx-th5qVJFH5ET:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L667)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

In light of the previous, here you'll want to iterate over `desc.PublicNonPrimaryIndexes()`.
</blockquote></details>

Done.
___
*[pkg/sql/alter_primary_key.go, line 670 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhOFzKAaslAs6U5riF:-Myh_G-X-Ww9hf7LZrra:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L670)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

You should also check that `existingIndex` is not a partial index. It isn't the most obvious thing in the world but we allow WHERE clauses in indexes! To reuse your example in the logic test, this could be something like `CREATE TABLE t (a INT PRIMARY KEY, b INT NOT NULL, UNIQUE INDEX t_a_key (a) WHERE (a > 0));`.

You can check for the presence of this partial predicate with the `IsPartial` method on `catalog.Index`.
</blockquote></details>

Done.
___
*[pkg/sql/alter_primary_key.go, line 674 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhOTTk5rTkzIOpyq3m:-MyhbLvK3T0P5rKdyaa3:bps68l2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L674)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I'm not sure if column directions matter here, or elsewhere in `shouldCopyPrimaryKey` for that matter. Let's ask @ajwerner 
</blockquote></details>

The reason I check column directions is one of the existing logic tests has an expected output that created a new secondary index for `ALTER PRIMARY KEY` while there is already a secondary index on the (old) primary key columns, but opposite direction.

If directions do not matter, then I'll need to modify existing logic tests. If so, is there any other guidelines I need to follow?
___
*[pkg/sql/alter_primary_key.go, line 678 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhOnr82RMSA560QotY:-MyhcyTT9FUKzxXlI2Hz:bk8hi9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L678)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

`catalog.Index` has a nice `CollectKeyColumnIDs` method which returns a `catalog.TableColSet` which should be much more convenient for what you want to do here, which is to test whether there exists a non-partial unique secondary index whose set of key columns is a superset of the PK columns.
</blockquote></details>

Currently, I code index equality as "same keyed columns, same size, in the same order". 

I think superset is too "loose" because of the unorderedness -- e.g. it's hard to justify not to create a secondary index on the (old) PK column "a" when there is an existing secondary index on ("b", "a"), since this existing index cannot be used in queries filtered by "a" (e.g.  SELECT ... WHERE "a" > 4)

I'd argue that a strict prefix will be a better index-equality definition. That is, we don't create a new secondary index on the (old) PK columns if there exists a secondary index s.t. (old) PK columns is a strict prefix of this secondary index columns.
This definition corresponds exactly to the additional test you suggested below.
However, such a change will also break one of the existing logic tests as expected. I've modified the existing logic test to accommodate this new code change.
___
*[pkg/sql/logictest/testdata/logic_test/alter_primary_key, line 1607 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhKb7m2mJQf3D_gCKi:-MyhgOXB7AoATJD353G8:b-pkn7uo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L1607)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This is a good test. Perhaps you can also add a case for `CREATE TABLE t (a INT PRIMARY KEY, b INT NOT NULL, UNIQUE INDEX idx (a, b));`, we should see a similar outcome: only 2 indexes (the old primary index and `idx`), the old primary index should be discarded.
</blockquote></details>

Added. Thank you for the suggestion.

**Review by postamar — COMMENTED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78046)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @Xiang-Gu)

___
*[-- commits, line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MysFp6W0FxAXj0t9Jkl):*
nit: I believe git commit message lines should not exceed a certain length, which is probably described in the style guide. Try to set up your editor to enforce this limit, that way you won't have to concern yourself with this ever again. This isn't strictly enforced by the way, so no big deal.
___
*[pkg/sql/alter_primary_key.go, line 678 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhOnr82RMSA560QotY:-MysGnZL5q9ZumV9lK0d:bd2airf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L678)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

Currently, I code index equality as "same keyed columns, same size, in the same order". 

I think superset is too "loose" because of the unorderedness -- e.g. it's hard to justify not to create a secondary index on the (old) PK column "a" when there is an existing secondary index on ("b", "a"), since this existing index cannot be used in queries filtered by "a" (e.g.  SELECT ... WHERE "a" > 4)

I'd argue that a strict prefix will be a better index-equality definition. That is, we don't create a new secondary index on the (old) PK columns if there exists a secondary index s.t. (old) PK columns is a strict prefix of this secondary index columns.
This definition corresponds exactly to the additional test you suggested below.
However, such a change will also break one of the existing logic tests as expected. I've modified the existing logic test to accommodate this new code change.
</blockquote></details>

My understanding, which may very well be wrong, was that the old primary index was guaranteeing a uniqueness property on the old primary key, and this property should be maintained even after replacing the old primary index with a new one. Your argument certainly has merit, though. @ajwerner can you weigh in here?

In any case this change of yours is a strict improvement to the existing code and so this question should not block merging this PR.
___
*[pkg/sql/logictest/testdata/logic_test/alter_primary_key, line 1208 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MysJhAd2d5xMMyO-PWR:-MysJhAd2d5xMMyO-PWS:bgz45cm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e25031ae7a45c02d3443b00ea4e55538b8f9b172/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L1208)):*
This is surprising, shouldn't this index be preserved, following your reasoning?

> Currently, I code index equality as "same keyed columns, same size, in the same order".

Reading your code I'm also not quite how this is happening. Am I missing something?

**Review by Xiang-Gu — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78046)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[-- commits, line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MysFp6W0FxAXj0t9Jkl):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: I believe git commit message lines should not exceed a certain length, which is probably described in the style guide. Try to set up your editor to enforce this limit, that way you won't have to concern yourself with this ever again. This isn't strictly enforced by the way, so no big deal.
</blockquote></details>

Done. 
___
*[pkg/sql/alter_primary_key.go, line 678 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhOnr82RMSA560QotY:-Mz1KrXGBBFQ4abYBjwY:b7ifyl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L678)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

My understanding, which may very well be wrong, was that the old primary index was guaranteeing a uniqueness property on the old primary key, and this property should be maintained even after replacing the old primary index with a new one. Your argument certainly has merit, though. @ajwerner can you weigh in here?

In any case this change of yours is a strict improvement to the existing code and so this question should not block merging this PR.
</blockquote></details>

`the old primary index was guaranteeing a uniqueness property on the old primary key, and this property should be maintained even after replacing the old primary index with a new one`
If we'd like to continue to maintain the uniqueness property on the old primary key column(s), then yes, we do need a "strict" equality definition (i.e. same column(s), same direction, in same order, unique, non-partial). I am just curious why we want to do that, so yeah, Andrew, please give us more insight here.

`In any case this change of yours is a strict improvement to the existing code and so this question should not block merging this PR.`
In the most recent change, I change the equality definition to strict-prefix (and thus also modified some existing logic tests). We do need to wait until we figure this out before merging.
___
*[pkg/sql/logictest/testdata/logic_test/alter_primary_key, line 1208 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MysJhAd2d5xMMyO-PWR:-Mz1J5Y22GuIQPH7tUCl:b-ocwqzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e25031ae7a45c02d3443b00ea4e55538b8f9b172/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L1208)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This is surprising, shouldn't this index be preserved, following your reasoning?

> Currently, I code index equality as "same keyed columns, same size, in the same order".

Reading your code I'm also not quite how this is happening. Am I missing something?

</blockquote></details>

In the most recent change of this PR, I changed the index-equality from "same keyed columns, same size, in the same order" to "strict-prefix".

This breaks some existing logic tests, such as this one, because we already have a secondary index (id, id2), and the old PK (id) is a strict prefix of it.

This is also the reason for other logic tests modification.

**Review by ajwerner — COMMENTED** [review]
I think I'd let the principle of lease surprise guide me here. Two unique indexes of the same exact columns would be surprising. I'm not sure much else would be. This is all an edge case for a behavior cockroachdb made up, so I more or less agree we can't mess up too badly here. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-:-MzIpmXf4YicsGXraY6R:b3ikrro)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar and @Xiang-Gu)

___
*[pkg/sql/alter_primary_key.go, line 674 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhOTTk5rTkzIOpyq3m:-MzIliQeD9OqyFFbzyH2:b-h9fdlw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L674)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

The reason I check column directions is one of the existing logic tests has an expected output that created a new secondary index for `ALTER PRIMARY KEY` while there is already a secondary index on the (old) primary key columns, but opposite direction.

If directions do not matter, then I'll need to modify existing logic tests. If so, is there any other guidelines I need to follow?
</blockquote></details>

This is all a convenience feature which postgres doesn't have. We can do whatever we want here. The reason we did this was out of fear that customers will experience very bad performance on queries they expected previously to work well. The difference in directions _can_ affect some queries, but is much less likely to than the lack of the index. Either choice would be valid. 
___
*[pkg/sql/alter_primary_key.go, line 678 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhOnr82RMSA560QotY:-MzInDfm3GXq9XK8uvOZ:b1hbn2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L678)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

`the old primary index was guaranteeing a uniqueness property on the old primary key, and this property should be maintained even after replacing the old primary index with a new one`
If we'd like to continue to maintain the uniqueness property on the old primary key column(s), then yes, we do need a "strict" equality definition (i.e. same column(s), same direction, in same order, unique, non-partial). I am just curious why we want to do that, so yeah, Andrew, please give us more insight here.

`In any case this change of yours is a strict improvement to the existing code and so this question should not block merging this PR.`
In the most recent change, I change the equality definition to strict-prefix (and thus also modified some existing logic tests). We do need to wait until we figure this out before merging.
</blockquote></details>

I think I agree that the order of the columns matters. I don't know how much we care about the direction. 
___
*[pkg/sql/alter_primary_key.go, line 631 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MzIpdmD34FXu8Ux_3DS:-MzIpdmD34FXu8Ux_3DT:b-61xr5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/8aa0f6c5826efcb68a8346421ce9b32bd2cadb1a/pkg/sql/alter_primary_key.go#L631)):*
> ```Go
> //   other than hash sharding.
> // * There is no partitioning change.
> // * There is no existing secondary index on the old primary key columns.
> ```

This is stale given the strict prefix change. Let's make sure the design aligns with the commentary before we merge. 
___
*[pkg/sql/alter_primary_key.go, line 670 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MzImzr7BX9agZNvTrNI:-MzImzr7BX9agZNvTrNJ:bkmy8k2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8aa0f6c5826efcb68a8346421ce9b32bd2cadb1a/pkg/sql/alter_primary_key.go#L670)):*
> ```Go
> 
> 		for _, existingIndex := range desc.PublicNonPrimaryIndexes() {
> 			if existingIndex.IsPartial() || !existingIndex.IsUnique() || pk.NumKeyColumns() > existingIndex.NumKeyColumns() {
> ```

nit: I encourage you to refactor this logic into yet another helper which takes two indexes and compares them on whatever properties we deem important. That way you can just early return a false. It'll also clean up the readability a tad. 



_Code quote:_
```Go
			if existingIndex.IsPartial() || !existingIndex.IsUnique() || pk.NumKeyColumns() > existingIndex.NumKeyColumns() {
				continue
			}
			isDifferent := false
```
___
*[pkg/sql/alter_primary_key.go, line 670 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MzInIqQA7SgWS3H-A38:-MzInIqQA7SgWS3H-A39:bi4vfv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8aa0f6c5826efcb68a8346421ce9b32bd2cadb1a/pkg/sql/alter_primary_key.go#L670)):*
> ```Go
> 
> 		for _, existingIndex := range desc.PublicNonPrimaryIndexes() {
> 			if existingIndex.IsPartial() || !existingIndex.IsUnique() || pk.NumKeyColumns() > existingIndex.NumKeyColumns() {
> ```

I'm not so sure about this strict prefix choice. At the very least it should be well commented. I do buy that it makes sense from a query performance perspective. One thing that you lose which is surprising is that we'll no longer be maintaining uniqueness over the same tuple. I also don't expect this case to come up very frequently, so it's not clear that it's worth the complexity when compared to . It'd be weird to have a unique secondary index over a tuple with the same columns in the same order with the same direction as the primary key, but with even more columns.

**Review by Xiang-Gu — COMMENTED** [review]
Thank you for the guidance. yeah, I will revert back to the definition of index equality to "same columns, same order, same direction" and consider only "unique, non-partial" existing secondary indexes.

RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-:-MzQFdzbElyj1VPaqmwY:befdlo8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/sql/alter_primary_key.go, line 674 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhOTTk5rTkzIOpyq3m:-MzQGYMt7h6TktWJMGIr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L674)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This is all a convenience feature which postgres doesn't have. We can do whatever we want here. The reason we did this was out of fear that customers will experience very bad performance on queries they expected previously to work well. The difference in directions _can_ affect some queries, but is much less likely to than the lack of the index. Either choice would be valid. 
</blockquote></details>

Done.
___
*[pkg/sql/alter_primary_key.go, line 678 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MyhOnr82RMSA560QotY:-MzQGst04xF90XvVeWeS:b-t9jkol) ([raw file](https://github.com/cockroachdb/cockroach/blob/b905629d29ef4a3041204514e5a018bd30f446a0/pkg/sql/alter_primary_key.go#L678)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think I agree that the order of the columns matters. I don't know how much we care about the direction. 
</blockquote></details>

acked.
___
*[pkg/sql/alter_primary_key.go, line 670 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MzImzr7BX9agZNvTrNI:-MzQGOol4U4JJtbu-_ob:b-7rl0kd) ([raw file](https://github.com/cockroachdb/cockroach/blob/8aa0f6c5826efcb68a8346421ce9b32bd2cadb1a/pkg/sql/alter_primary_key.go#L670)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: I encourage you to refactor this logic into yet another helper which takes two indexes and compares them on whatever properties we deem important. That way you can just early return a false. It'll also clean up the readability a tad. 
</blockquote></details>

done
___
*[pkg/sql/alter_primary_key.go, line 670 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MzInIqQA7SgWS3H-A38:-MzQGSD297LF_F-aT99S:b-y4hucb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8aa0f6c5826efcb68a8346421ce9b32bd2cadb1a/pkg/sql/alter_primary_key.go#L670)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm not so sure about this strict prefix choice. At the very least it should be well commented. I do buy that it makes sense from a query performance perspective. One thing that you lose which is surprising is that we'll no longer be maintaining uniqueness over the same tuple. I also don't expect this case to come up very frequently, so it's not clear that it's worth the complexity when compared to . It'd be weird to have a unique secondary index over a tuple with the same columns in the same order with the same direction as the primary key, but with even more columns. 
</blockquote></details>

acked. reverted back to strict-equality definition.

**Review by ajwerner — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-:-MzaZuDJ6-OLBlXS0jLT:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @postamar, and @Xiang-Gu)

___
*[pkg/sql/alter_primary_key.go, line 681 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MzaZVV2DDBTO5ajo5y2:-MzaZVV2DDBTO5ajo5y3:b-7ftb4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f2c7865f96fbcf6b016a8b34b6adafe164dc679/pkg/sql/alter_primary_key.go#L681)):*
> ```Go
> 		isIdentical := func(idx1, idx2 catalog.Index) bool {
> 			if idx1.IsUnique() != idx2.IsUnique() ||
> 				idx1.IsPartial() != idx2.IsPartial() ||
> ```

I think that if either of them are partial, you would want to check if they have the same predicate. In practice, we know the primary index won't be partial, so :shrug: , but it feels weird to think this is enough to say that they are identical. Somebody might mindlessly refactor this in the future thinking it's complete.

**Review by Xiang-Gu — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78046)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/sql/alter_primary_key.go, line 681 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/78046#-MzaZVV2DDBTO5ajo5y2:-MzpW4-oB9q8-I7VI0OK:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f2c7865f96fbcf6b016a8b34b6adafe164dc679/pkg/sql/alter_primary_key.go#L681)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think that if either of them are partial, you would want to check if they have the same predicate. In practice, we know the primary index won't be partial, so :shrug: , but it feels weird to think this is enough to say that they are identical. Somebody might mindlessly refactor this in the future thinking it's complete. 
</blockquote></details>

done

**Pull request merged**
